File: src/shortcut.test.ts
import { Identifier } from "@tozd/identifier"
import { assert, describe, expect, test } from "vitest"

import { createShortcutToQuery, parseShortcut, resolveShortcutID, shortcutToFilters, shortcutToQuery } from "@/shortcut"

describe("parseShortcut", () => {
  test("parses a single key/value pair", () => {
    assert.deepEqual(parseShortcut("ns.example.com,KIND=ns.example.com,A"), [{ key: "ns.example.com,KIND", value: "ns.example.com,A" }])
  })

  test("parses multiple parts separated by &", () => {
    assert.deepEqual(parseShortcut("a=1,2&b:c=3,4"), [
      { key: "a", value: "1,2" },
      { key: "b:c", value: "3,4" },
    ])
  })

  test("preserves only the first '=' as separator", () => {
    assert.deepEqual(parseShortcut("key=value=with=equals"), [{ key: "key", value: "value=with=equals" }])
  })

  test("preserves 'self' value literally", () => {
    assert.deepEqual(parseShortcut("ns.example.com,KIND=self"), [{ key: "ns.example.com,KIND", value: "self" }])
  })

  test("preserves 'reverse' key literally", () => {
    assert.deepEqual(parseShortcut("reverse=ns.example.com,DOC"), [{ key: "reverse", value: "ns.example.com,DOC" }])
  })

  test("throws on empty input", () => {
    expect(() => parseShortcut("")).toThrowError("search shortcut must not be empty")
  })

  test("throws when '=' is missing", () => {
    expect(() => parseShortcut("ns.example.com,KIND")).toThrowError(/non-empty key and value/)
  })

  test("throws on empty value", () => {
    expect(() => parseShortcut("ns.example.com,KIND=")).toThrowError(/non-empty key and value/)
  })

  test("throws when key has more than one ':'", () => {
    expect(() => parseShortcut("a:b:c=ns.example.com,D")).toThrowError(/at most one ':'/)
  })

  test("throws when 'reverse' is the parent of a nested key", () => {
    expect(() => parseShortcut("reverse:ns.example.com,X=ns.example.com,Y")).toThrowError(/"reverse" is not allowed/)
  })

  test("throws when 'reverse' is the nested side of a key", () => {
    expect(() => parseShortcut("ns.example.com,X:reverse=ns.example.com,Y")).toThrowError(/"reverse" is not allowed/)
  })
})

describe("resolveShortcutID", () => {
  test("hashes comma-separated parts via Identifier.from", async () => {
    const want = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.equal(await resolveShortcutID("ns.example.com,KIND"), want)
  })

  test("returns single tokens unchanged", async () => {
    const id = Identifier.new().toString()
    assert.equal(await resolveShortcutID(id), id)
  })
})

describe("shortcutToFilters", () => {
  test("builds a ref filter from a multi-part key and value", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=ns.example.com,A")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const value = (await Identifier.from("ns.example.com", "A")).toString()
    assert.deepEqual(payload, {
      filters: [{ prop: [prop], ref: { to: [{ id: value }] } }],
    })
  })

  test("builds a nested ref filter for 'parent:prop' keys", async () => {
    const payload = await shortcutToFilters("ns.example.com,P:ns.example.com,Q=ns.example.com,V")
    const parent = (await Identifier.from("ns.example.com", "P")).toString()
    const nested = (await Identifier.from("ns.example.com", "Q")).toString()
    const value = (await Identifier.from("ns.example.com", "V")).toString()
    assert.deepEqual(payload.filters, [{ prop: [parent, nested], ref: { to: [{ id: value }] } }])
  })

  test("groups repeated keys into one filter's 'to' list", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=ns.example.com,A&ns.example.com,KIND=ns.example.com,B")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const a = (await Identifier.from("ns.example.com", "A")).toString()
    const b = (await Identifier.from("ns.example.com", "B")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { to: [{ id: a }, { id: b }] } }])
  })

  test("keeps distinct properties as separate filters in first-appearance order", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=ns.example.com,A&ns.example.com,OTHER=ns.example.com,B&ns.example.com,KIND=ns.example.com,C")
    const kind = (await Identifier.from("ns.example.com", "KIND")).toString()
    const other = (await Identifier.from("ns.example.com", "OTHER")).toString()
    const a = (await Identifier.from("ns.example.com", "A")).toString()
    const b = (await Identifier.from("ns.example.com", "B")).toString()
    const c = (await Identifier.from("ns.example.com", "C")).toString()
    assert.deepEqual(payload.filters, [
      { prop: [kind], ref: { to: [{ id: a }, { id: c }] } },
      { prop: [other], ref: { to: [{ id: b }] } },
    ])
  })

  test("substitutes 'self' with the supplied self ID", async () => {
    const self = Identifier.new().toString()
    const payload = await shortcutToFilters("ns.example.com,KIND=self", self)
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { to: [{ id: self }] } }])
  })

  test("throws when 'self' is referenced without a self prop", async () => {
    await expect(shortcutToFilters("ns.example.com,KIND=self")).rejects.toThrowError(/no self ID was provided/)
  })

  test("sets reverse at the top level", async () => {
    const payload = await shortcutToFilters("reverse=ns.example.com,DOC")
    const value = (await Identifier.from("ns.example.com", "DOC")).toString()
    assert.deepEqual(payload, { reverse: value })
  })

  test("supports reverse with self", async () => {
    const self = Identifier.new().toString()
    const payload = await shortcutToFilters("reverse=self", self)
    assert.deepEqual(payload, { reverse: self })
  })

  test("combines reverse and filters in one payload", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=ns.example.com,A&reverse=ns.example.com,DOC")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const value = (await Identifier.from("ns.example.com", "A")).toString()
    const rev = (await Identifier.from("ns.example.com", "DOC")).toString()
    assert.deepEqual(payload, {
      reverse: rev,
      filters: [{ prop: [prop], ref: { to: [{ id: value }] } }],
    })
  })

  test("omits filters when only reverse is present", async () => {
    const payload = await shortcutToFilters("reverse=ns.example.com,DOC")
    assert.notProperty(payload, "filters")
  })

  test("builds a missing selection from a 'missing' value", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=missing")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { missing: true } }])
  })

  test("builds a direct selection from a 'direct:' value", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=direct:ns.example.com,A")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const value = (await Identifier.from("ns.example.com", "A")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { direct: [{ id: value }] } }])
  })

  test("groups to, direct, and missing for the same property into one filter", async () => {
    const payload = await shortcutToFilters("ns.example.com,KIND=ns.example.com,A&ns.example.com,KIND=direct:ns.example.com,B&ns.example.com,KIND=missing")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const a = (await Identifier.from("ns.example.com", "A")).toString()
    const b = (await Identifier.from("ns.example.com", "B")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { to: [{ id: a }], direct: [{ id: b }], missing: true } }])
  })

  test("substitutes 'self' inside a direct value", async () => {
    const self = Identifier.new().toString()
    const payload = await shortcutToFilters("ns.example.com,KIND=direct:self", self)
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.deepEqual(payload.filters, [{ prop: [prop], ref: { direct: [{ id: self }] } }])
  })
})

describe("shortcutToQuery", () => {
  test("emits a single key with a one-element list of resolved identifiers", async () => {
    const query = await shortcutToQuery("ns.example.com,KIND=ns.example.com,A")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const value = (await Identifier.from("ns.example.com", "A")).toString()
    assert.deepEqual(query, { [prop]: [value] })
  })

  test("joins nested keys with ':'", async () => {
    const query = await shortcutToQuery("ns.example.com,P:ns.example.com,Q=ns.example.com,V")
    const parent = (await Identifier.from("ns.example.com", "P")).toString()
    const nested = (await Identifier.from("ns.example.com", "Q")).toString()
    const value = (await Identifier.from("ns.example.com", "V")).toString()
    assert.deepEqual(query, { [`${parent}:${nested}`]: [value] })
  })

  test("groups repeated keys into a list of values", async () => {
    const query = await shortcutToQuery("ns.example.com,KIND=ns.example.com,A&ns.example.com,KIND=ns.example.com,B")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const a = (await Identifier.from("ns.example.com", "A")).toString()
    const b = (await Identifier.from("ns.example.com", "B")).toString()
    assert.deepEqual(query, { [prop]: [a, b] })
  })

  test("preserves 'reverse' as the literal key", async () => {
    const query = await shortcutToQuery("reverse=ns.example.com,DOC")
    const value = (await Identifier.from("ns.example.com", "DOC")).toString()
    assert.deepEqual(query, { reverse: [value] })
  })

  test("substitutes 'self' with the supplied self ID", async () => {
    const self = Identifier.new().toString()
    const query = await shortcutToQuery("ns.example.com,KIND=self", self)
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.deepEqual(query, { [prop]: [self] })
  })

  test("throws when 'self' is referenced without a self prop", async () => {
    await expect(shortcutToQuery("ns.example.com,KIND=self")).rejects.toThrowError(/no self ID was provided/)
  })

  test("encodes a missing selection as the 'missing' value", async () => {
    const query = await shortcutToQuery("ns.example.com,KIND=missing")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    assert.deepEqual(query, { [prop]: ["missing"] })
  })

  test("encodes a direct selection with the 'direct:' prefix", async () => {
    const query = await shortcutToQuery("ns.example.com,KIND=direct:ns.example.com,A")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const value = (await Identifier.from("ns.example.com", "A")).toString()
    assert.deepEqual(query, { [prop]: ["direct:" + value] })
  })

  test("groups to, direct, and missing values under one key", async () => {
    const query = await shortcutToQuery("ns.example.com,KIND=ns.example.com,A&ns.example.com,KIND=direct:ns.example.com,B&ns.example.com,KIND=missing")
    const prop = (await Identifier.from("ns.example.com", "KIND")).toString()
    const a = (await Identifier.from("ns.example.com", "A")).toString()
    const b = (await Identifier.from("ns.example.com", "B")).toString()
    assert.deepEqual(query, { [prop]: [a, "direct:" + b, "missing"] })
  })
})

describe("createShortcutToQuery", () => {
  test("keeps 'limit' as a literal key and resolves its value", async () => {
    const query = await createShortcutToQuery("limit=ns.example.com,RESOURCE")
    const resource = (await Identifier.from("ns.example.com", "RESOURCE")).toString()
    assert.deepEqual(query, { limit: [resource] })
  })

  test("resolves a property=value entry into an initial reference claim", async () => {
    const query = await createShortcutToQuery("ns.example.com,PROP=ns.example.com,VAL")
    const prop = (await Identifier.from("ns.example.com", "PROP")).toString()
    const value = (await Identifier.from("ns.example.com", "VAL")).toString()
    assert.deepEqual(query, { [prop]: [value] })
  })

  test("substitutes 'self' in the value with the supplied self ID", async () => {
    const self = Identifier.new().toString()
    const query = await createShortcutToQuery("ns.example.com,PROP=self", self)
    const prop = (await Identifier.from("ns.example.com", "PROP")).toString()
    assert.deepEqual(query, { [prop]: [self] })
  })

  test("combines a limit with a self-referencing property claim", async () => {
    const self = Identifier.new().toString()
    const query = await createShortcutToQuery("limit=ns.example.com,RESOURCE&ns.example.com,PROP=self", self)
    const resource = (await Identifier.from("ns.example.com", "RESOURCE")).toString()
    const prop = (await Identifier.from("ns.example.com", "PROP")).toString()
    assert.deepEqual(query, { limit: [resource], [prop]: [self] })
  })

  test("throws when 'self' is referenced without a self ID", async () => {
    await expect(createShortcutToQuery("ns.example.com,PROP=self")).rejects.toThrowError(/no self ID was provided/)
  })
})

File: transform/shortcut_test.go
package transform_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"gitlab.com/tozd/identifier"

	"gitlab.com/peerdb/peerdb/transform"
)

//nolint:maintidx
func TestValidateShortcut(t *testing.T) {
	t.Parallel()

	t.Run("multi-part key and value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("core.namespace,INSTANCE_OF=ns.example.com,A")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("multiple parts with &", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=ns.example.com,OPT_A&ns.example.com,OTHER=ns.example.com,OPT_B")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("nested key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,LOCATED_IN:ns.example.com,COUNTRY=ns.example.com,FRANCE")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("base58 identifier in key and value", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut(id + "=" + id)
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("base58 identifier in value with multi-part key", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("ns.example.com,KIND=" + id)
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("base58 identifier in key with multi-part value", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut(id + "=ns.example.com,A")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("base58 identifier in nested key", func(t *testing.T) {
		t.Parallel()

		parent := identifier.New().String()
		prop := identifier.New().String()
		errE := transform.TestingValidateShortcut(parent + ":" + prop + "=ns.example.com,A")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("base58 identifier in reverse value", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("reverse=" + id)
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("reverse key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("reverse=ns.example.com,DOC")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("self value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=self")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("missing value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=missing")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("missing value in nested key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,LOCATED_IN:ns.example.com,COUNTRY=missing")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("direct value with multi-part identifier", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=direct:ns.example.com,A")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("direct value with base58 identifier", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("ns.example.com,KIND=direct:" + id)
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("direct value in nested key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,LOCATED_IN:ns.example.com,COUNTRY=direct:ns.example.com,FRANCE")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("direct self value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=direct:self")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("id key with base58 identifier", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("id=" + id)
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("id key with multi-part identifier", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=ns.example.com,DOC")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("id key with self value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=self")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("id key with languages value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=languages")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("languages value is rejected outside the id key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("reverse=languages")
		require.Error(t, errE)
	})

	t.Run("repeated id key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=ns.example.com,A&id=ns.example.com,B")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("id key with property values", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=ns.example.com,A&ns.example.com,KIND=ns.example.com,OPT_A")
		require.NoError(t, errE, "% -+#.1v", errE)
	})

	t.Run("reverse with missing value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("reverse=missing")
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut value is not a valid identifier")
	})

	t.Run("reverse with direct value", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("reverse=direct:" + id)
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut value must be a single identifier")
	})

	t.Run("id with missing value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id=missing")
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut value is not a valid identifier")
	})

	t.Run("id with direct value", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("id=direct:" + id)
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut value must be a single identifier")
	})

	t.Run("multi-segment value not starting with direct", func(t *testing.T) {
		t.Parallel()

		id := identifier.New().String()
		errE := transform.TestingValidateShortcut("ns.example.com,KIND=other:" + id)
		require.Error(t, errE)
		assert.EqualError(t, errE, `search shortcut multi-segment value must start with "direct"`)
	})

	t.Run("direct value with invalid identifier", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=direct:bogus")
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut direct value is not a valid identifier")
	})

	t.Run("value with too many colons", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=a:b:c")
		require.Error(t, errE)
		assert.EqualError(t, errE, `search shortcut value must contain at most one ":"`)
	})

	t.Run("empty", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("")
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut must not be empty")
	})

	t.Run("missing equals", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND")
		require.Error(t, errE)
		assert.EqualError(t, errE, "entry must have a non-empty key and value separated by '='")
	})

	t.Run("empty value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=")
		require.Error(t, errE)
		assert.EqualError(t, errE, "entry must have a non-empty key and value separated by '='")
	})

	t.Run("key with too many colons", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("a:b:c=ns.example.com,D")
		require.Error(t, errE)
		assert.EqualError(t, errE, "search shortcut key must contain at most one ':'")
	})

	t.Run("reverse inside nested key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("reverse:ns.example.com,X=ns.example.com,Y")
		require.Error(t, errE)
		assert.EqualError(t, errE, `"reverse" is not allowed inside a nested key`)
	})

	t.Run("id inside nested key", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("id:ns.example.com,X=ns.example.com,Y")
		require.Error(t, errE)
		assert.EqualError(t, errE, `"id" is not allowed inside a nested key`)
	})

	t.Run("invalid identifier in value", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=bogus")
		require.Error(t, errE)
	})

	t.Run("empty part inside multi-part identifier", func(t *testing.T) {
		t.Parallel()

		errE := transform.TestingValidateShortcut("ns.example.com,KIND=ns.example.com,,A")
		require.Error(t, errE)
		assert.EqualError(t, errE, "empty identifier part")
	})
}
